### ISSUE TITLE
pods/log: NotFound resource

### ISSUE THREAD

**Issue by xtuc** [kubernetes/kubernetes#60835]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:

I'm not able to request the logs of my pods from non master nodes.

Example:

```sh
$ kubectl logs redis-deployment-5697dff764-k6pns -v=9

I0306 09:18:36.559814   24182 loader.go:357] Config loaded from file /etc/kubernetes/kubelet.conf
I0306 09:18:36.581218   24182 round_trippers.go:417] curl -k -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: kubectl/v1.9.3 (linux/amd64) kubernetes/d283541" https://192.168.99.2:6443/api/v1/namespaces/default/pods/redis-deployment-5697dff764-k6pns
I0306 09:18:36.589393   24182 round_trippers.go:436] GET https://192.168.99.2:6443/api/v1/namespaces/default/pods/redis-deployment-5697dff764-k6pns 200 OK in 7 milliseconds
I0306 09:18:36.589735   24182 round_trippers.go:442] Response Headers:
I0306 09:18:36.589956   24182 round_trippers.go:445]     Content-Type: application/json
I0306 09:18:36.590163   24182 round_trippers.go:445]     Content-Length: 2461
I0306 09:18:36.590359   24182 round_trippers.go:445]     Date: Tue, 06 Mar 2018 09:18:36 GMT
I0306 09:18:36.590758   24182 request.go:873] Response Body: {"kind":"Pod","apiVersion":"v1","metadata":{"name":"redis-deployment-5697dff764-k6pns","generateName":"redis-deployment-5697dff764-","namespace":"default","selfLink":"/api/v1/namespaces/default/pods/redis-deployment-5697dff764-k6pns","uid":"673e0953-2112-11e8-a5d9-0238074c8285","resourceVersion":"885","creationTimestamp":"2018-03-06T07:45:56Z","labels":{"app":"redis","pod-template-hash":"1253899320"},"ownerReferences":[{"apiVersion":"extensions/v1beta1","kind":"ReplicaSet","name":"redis-deployment-5697dff764","uid":"673cc513-2112-11e8-a5d9-0238074c8285","controller":true,"blockOwnerDeletion":true}]},"spec":{"volumes":[{"name":"volume","hostPath":{"path":"/data/redis","type":"DirectoryOrCreate"}},{"name":"default-token-nxvgw","secret":{"secretName":"default-token-nxvgw","defaultMode":420}}],"containers":[{"name":"redis","image":"redis","ports":[{"containerPort":6379,"protocol":"TCP"}],"resources":{},"volumeMounts":[{"name":"volume","mountPath":"/data"},{"name":"default-token-nxvgw","readOnly":true,"mountPath":"/var/run/secrets/kubernetes.io/serviceaccount"}],"terminationMessagePath":"/dev/termination-log","terminationMessagePolicy":"File","imagePullPolicy":"Always"}],"restartPolicy":"Always","terminationGracePeriodSeconds":30,"dnsPolicy":"ClusterFirst","serviceAccountName":"default","serviceAccount":"default","nodeName":"node-192-168-99-3","securityContext":{},"schedulerName":"default-scheduler","tolerations":[{"key":"node.kubernetes.io/not-ready","operator":"Exists","effect":"NoExecute","tolerationSeconds":300},{"key":"node.kubernetes.io/unreachable","operator":"Exists","effect":"NoExecute","tolerationSeconds":300}]},"status":{"phase":"Running","conditions":[{"type":"Initialized","status":"True","lastProbeTime":null,"lastTransitionTime":"2018-03-06T07:47:00Z"},{"type":"Ready","status":"True","lastProbeTime":null,"lastTransitionTime":"2018-03-06T07:47:06Z"},{"type":"PodScheduled","status":"True","lastProbeTime":null,"lastTransitionTime":"2018-03-06T07:46:59Z"}],"hostIP":"10.0.2.15","podIP":"10.67.1.163","startTime":"2018-03-06T07:47:00Z","containerStatuses":[{"name":"redis","state":{"running":{"startedAt":"2018-03-06T07:47:06Z"}},"lastState":{},"ready":true,"restartCount":0,"image":"redis:latest","imageID":"docker-pullable://redis@sha256:0bf55b655c461a6f41a7c83c407e1e90be9dbb9aa41831e93cc1a96621530669","containerID":"docker://4e0d97e508b57252cdd18e388b06f20e54d6995ed2a852290eea7290d890293d"}],"qosClass":"BestEffort"}}
I0306 09:18:36.667654   24182 round_trippers.go:417] curl -k -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: kubectl/v1.9.3 (linux/amd64) kubernetes/d283541" https://192.168.99.2:6443/api/v1/namespaces/default/pods/redis-deployment-5697dff764-k6pns/log
I0306 09:18:36.671694   24182 round_trippers.go:436] GET https://192.168.99.2:6443/api/v1/namespaces/default/pods/redis-deployment-5697dff764-k6pns/log 404 Not Found in 4 milliseconds
I0306 09:18:36.671723   24182 round_trippers.go:442] Response Headers:
I0306 09:18:36.671730   24182 round_trippers.go:445]     Content-Length: 280
I0306 09:18:36.671739   24182 round_trippers.go:445]     Date: Tue, 06 Mar 2018 09:18:36 GMT
I0306 09:18:36.671745   24182 round_trippers.go:445]     Content-Type: application/json
I0306 09:18:36.671797   24182 request.go:873] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the server could not find the requested resource ( pods/log redis-deployment-5697dff764-k6pns)","reason":"NotFound","details":{"name":"redis-deployment-5697dff764-k6pns","kind":"pods/log"},"code":404}
I0306 09:18:36.672037   24182 helpers.go:201] server response object: [{
  "metadata": {},
  "status": "Failure",
  "message": "the server could not find the requested resource ( pods/log redis-deployment-5697dff764-k6pns)",
  "reason": "NotFound",
  "details": {
    "name": "redis-deployment-5697dff764-k6pns",
    "kind": "pods/log"
  },
  "code": 404
}]
F0306 09:18:36.672070   24182 helpers.go:119] Error from server (NotFound): the server could not find the requested resource ( pods/log redis-deployment-5697dff764-k6pns)
```

I don't see anything in the logs about that.

**What you expected to happen**:

The logs.

**How to reproduce it (as minimally and precisely as possible)**:

I bootstraped the cluster using kubeadm. I'm not sure how you could reproduce it, here's the config I use (no special config files):

Kubelet:
```sh
/usr/bin/kubelet --bootstrap-kubeconfig=/etc/kubernetes/bootstrap-kubelet.conf --kubeconfig=/etc/kubernetes/kubelet.conf --pod-manifest-path=/etc/kubernetes/manifests --allow-privileged=true --network-plugin=cni --cni-conf-dir=/etc/cni/net.d --cni-bin-dir=/opt/cni/bin --cluster-dns=10.67.0.4 --cluster-domain=cluster.local --authorization-mode=Webhook --client-ca-file=/etc/kubernetes/pki/ca.crt --cadvisor-port=0 --rotate-certificates=true --cert-dir=/var/lib/kubelet/pki
```

apiserver:
```sh
kube-apiserver --requestheader-group-headers=X-Remote-Group --advertise-address=192.168.99.2 --service-account-key-file=/etc/kubernetes/pki/sa.pub --secure-port=6443 --requestheader-client-ca-file=/etc/kubernetes/pki/front-proxy-ca.crt --proxy-client-cert-file=/etc/kubernetes/pki/front-proxy-client.crt --enable-bootstrap-token-auth=true --allow-privileged=true --requestheader-extra-headers-prefix=X-Remote-Extra- --requestheader-allowed-names=front-proxy-client --client-ca-file=/etc/kubernetes/pki/ca.crt --kubelet-client-key=/etc/kubernetes/pki/apiserver-kubelet-client.key --insecure-port=0 --admission-control=Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,ResourceQuota --tls-private-key-file=/etc/kubernetes/pki/apiserver.key --kubelet-client-certificate=/etc/kubernetes/pki/apiserver-kubelet-client.crt --service-cluster-ip-range=10.66.0.0/16 --tls-cert-file=/etc/kubernetes/pki/apiserver.crt --proxy-client-key-file=/etc/kubernetes/pki/front-proxy-client.key --kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname --requestheader-username-headers=X-Remote-User --authorization-mode=Node,RBAC --etcd-servers=http://127.0.0.1:2379
```

**Anything else we need to know?**:

Related https://github.com/kubernetes/kubernetes/issues/45911, https://github.com/kubernetes/kubernetes/issues/29480

The issue here https://github.com/kubernetes/kubernetes/issues/38178 but I tried with/without `--enable-debugging-handlers=true` which yields the same result.

**Environment**:
- Kubernetes version (use `kubectl version`): `v1.9.3`
- Cloud provider or hardware configuration: Vagrant
- OS (e.g. from /etc/os-release): `Ubuntu 16.04.3 LTS`
- Kernel (e.g. `uname -a`): `Linux master-192-168-99-2 4.4.0-112-generic #135-Ubuntu SMP Fri Jan 19 11:48:36 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools: `kubeadm`
- Others: `vim 7.4.1689`

Thanks!

**Comment by xtuc** [comment]
I have no idea if this is the correct group; @kubernetes/sig-api-machinery-bugs. Thanks

**Comment by xtuc** [comment]
Interestingly the same configuration seems to work on AWS.

**Environment**:
- Kubernetes version (use `kubectl version`): `v1.9.3`
- Cloud provider or hardware configuration: AWS
- OS (e.g. from /etc/os-release): `Ubuntu 16.04.3 LTS`
- Kernel (e.g. `uname -a`): `Linux master-172-31-92-194 4.4.0-1049-aws #58-Ubuntu SMP Fri Jan 12 23:17:09 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools: `kubeadm`
- Others: `vim 7.4.1689`

**Comment by lavalamp** [comment]
Are you saying it works from master nodes but not from regular nodes? That doesn't make very much sense to me, I'd expect it to work everywhere or not at all.

/sig node 

(just in case)

**Comment by xtuc** [comment]
> Are you saying it works from master nodes but not from regular nodes? 

Yes.

Since it works with the same(ish) config on AWS, that might not be a Kubernetes issue.

**Comment by myguddy** [comment]
I have the same problem.
kubectl get pods returns
-->
hello-world-deployment-7dc8b47cbc-9btgn

but running logs and exec like
kubectl logs hello-world-deployment-7dc8b47cbc-9btgn
Error from server (NotFound): the server could not find the requested resource ( pods/log hello-world-deployment-7dc8b47cbc-9btgn)

kubectl exec -it hello-world-deployment-7dc8b47cbc-9btgn -- bash
error: unable to upgrade connection: pod does not exist

My Environment is on mac
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.5", GitCommit:"f01a2bf98249a4db383560443a59bed0c13575df", GitTreeState:"clean", BuildDate:"2018-03-19T19:38:17Z", GoVersion:"go1.9.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.5", GitCommit:"f01a2bf98249a4db383560443a59bed0c13575df", GitTreeState:"clean", BuildDate:"2018-03-19T15:50:45Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

Any solution or the way to find out what is wrong my k8s?

**Comment by verizonold** [comment]
I have the same problem as above. Any solution?

**Comment by xtuc** [comment]
@myguddy @verizonold please fill the issue template with your environment. I mentioned that I only have this issue on Vagrant.

**Comment by kaizamm** [comment]
i have the same problem，has this problem solved?

**Comment by xtuc** [comment]
@kaizamm what's your platform? This issue is on Vagrant

**Comment by kaizamm** [comment]
yeah, my plantform is vagrantl .  is this issue solved?

**Comment by verizonold** [comment]
I am using vagrant only for VM spin up. I deploy kubernetes using ansible and start up the master and minions manually. When I do a "kubectl apply...", I do see my pod deployed and Running. However, when I try to do a "kubectl exec..", I get:
kubectl exec -it ......   -- bash
error: unable to upgrade connection: pod does not exist

**Comment by kaizamm** [comment]
From flannel office website， i found that comment as follows：

Vagrant
Vagrant typically assigns two interfaces to all VMs. The first, for which all hosts are assigned the IP address 10.0.2.15, is for external traffic that gets NATed.

This may lead to problems with flannel. By default, flannel selects the first interface on a host. This leads to all hosts thinking they have the same public IP address. To prevent this issue, pass the --iface eth1 flag to flannel so that the second interface is chosen.


https://coreos.com/flannel/docs/latest/troubleshooting.html

**Comment by Arkadiusz234** [comment]
Have same problem with vagrant platform, any suggests how to resolve this problem ?

**Comment by jtopper** [comment]
This happens in Vagrant if you're running a multi-box setup because the kubelet on the workers end up binding services to the wrong ethernet interface. This error is the master attempting a connection to the wrong address in order to pull logs. The fix is to modify the config for the kubelet, as described in this blog post: https://medium.com/@joatmon08/playing-with-kubeadm-in-vagrant-machines-part-2-bac431095706

**Comment by geerlingguy** [comment]
Same thing here, and @jtopper is correct, it's because the first interface (in my case, with Debian 9 Stretch, `enp0s3`), is an internal interface used by Vagrant.

You have to use a secondary interface if using private IPs with Vagrant (again in my case, `enp0s8`).

Looking at the `logs` command with debug (`-v=9`), I found the pod was showing:

```
    "hostIP": "10.0.2.15",
    "podIP": "10.244.2.4",
```

And the hostIP corresponded to that virtual network interface that was internal to the VM... so effectively `kubectl` was querying localhost to see a pod that was on a different Vagrant-powered VM.

To summarize the linked post in the comment above, you have to edit the kubelet config (in `/etc/systemd/system/kubelet.service.d/10-kubeadm.conf`) and add or modify the `KUBELET_EXTRA_ARGS` to have the `--node-ip` flag, like so:

```
Environment="KUBELET_EXTRA_ARGS=--node-ip=VAGRANT_VM_EXTERNAL_IP_HERE"
```

Then restart `kubelet`.

**Comment by aario** [comment]
Bug still remains.
And solution by @geerlingguy didn't work:
```
$ sudo vi /etc/systemd/system/kubelet.service.d/10-kubeadm.conf
#Edited as described. Added IP of one of the worker nodes, then tried with ip of the other worker node
$ sudo systemctl daemon-reload
$ sudo systemctl restart kubelet
$ kubectl -n kube-system get pods  | grep hog
hog-6f56b7f4c4-k59rn                 1/1     Running            0          15m
$ kubectl -n kube-system logs hog-6f56b7f4c4-k59rn
Error from server (NotFound): the server could not find the requested resource ( pods/log hog-6f56b7f4c4-k59rn)
```

**Comment by orzroa** [comment]
@aario 
It works for me. Once you restarted kubelet, type "kubectl get pods -owide --all-namespaces", you will see some pods will change their IPs from 10.0.2.15 to another IP. After that "kubectl logs" works perfect.

**Comment by aario** [comment]
Ok. I found what was wrong in my case. The pod was created inside kube-system namespace. So the correct command was:
`kubectl -n kube-system logs <pod name i.e. hog-6848f56478-5gn9s>`
The trick is you cannot use `--all-namespaces` here. You must definitely mention the namespace in from of `-n` argument.

**Comment by qhh0205** [comment]
@geerlingguy‘s Solution worked for me.
